src/testing.py

```diff
@@ -89,6 +89,46 @@ def __call__(self, test_file):
       return Result(test=basename, success=False, output=e.output)
 
 
+def parse_exclude_files(fails, config_attributes):
+  ''' Parse the files containing tests to exclude (i.e. expected fails).
```

Comment:
This is kind of a block of text. How about a bulleted list, with nesting for the examples? I realize there is only so much to be done within 79 characters, but the interleaved parenthetical statements made it hard to read at first.

[PEP257](https://www.python.org/dev/peps/pep-0257/) nits:
* First line should be an executive summary, followed by a blank line
* Docstrings should use triple-doubles

Quick reformatting:
```
"""Returns a sorted list of exclusions which match the attributes.

Parse the files containing tests to exclude (i.e. expected fails).
* Each line may contain a comma-separated list of attributes restricting
  the test configurations which are expected to fail.
    - e.g. JS engine or optimization level.
* A test is only excluded if the configuration has all the attributes
  specified in the exclude line.
* Lines which have no attributes will match everything, and lines which
  specify only one attribute (e.g. engine) will match all configurations
  with that attribute (e.g. both opt levels with that engine).
"""
```

Given that, the last bullet feels like it can be reworded to split into two, which lets the parentheticals drop down more naturally.

Also given that we're many lines in at this point, some example lines might be useful